ACPs/103-dynamic-fees/README.md

```diff
@@ -0,0 +1,133 @@
+
+At the start of building/executing block $b$, $x$ is updated:
+
+$$x = \max(x - (T \cdot \Delta t), 0)$$
+
+Where $\Delta t$ is the number of seconds between $b$ and $b$'s parent block.
+
+The required fee per gas for block $b$ is:
+
+$$M \cdot \exp\left(\frac{x}{K}\right)$$
+
+Where:
+
+- $M$ is the minimum gas price
+- $\exp\left(x\right)$ is an approximation of $e^x$
+- $K$ is a constant to control the rate of change for gas price
+
+After processing block $b$, $x$ is updated with the total gas in the block $G$:
+
+$$x = x + G$$
+
+Whenever $x$ increases by $K$, the gas price increases by a factor of `~2.7`. If the gas price gets too expensive, average usage drops, and $x$ starts decreasing, automatically dropping the price again. The gas price constantly adjusts to make sure that, on average, the blockchain uses $T$ gas per second.
+
+$L$ is defined as the maximum gas per second that the blockchain has capacity for. For monotonically increasing blocktimes (as on P-Chain and X-Chain), there may be multiple blocks issued at the same on-chain timestamp. $L$ limits the amount of gas that can be used at $t$, placing a bound on the amount that the required fee can increase from $t$ to $t'$. Once the chain has used $L$ amount of gas at time $t$, a new block will only be considered valid if its timestamp is at least $t + 1$.
+
+The initial parameters for the X-Chain will be set to:
+
+| Parameter | Value |
+| - | - |
+| $M$ - minimum gas price | TODO |
+| $T$ - target gas per second | TODO |
+| $L$ - max gas per second | TODO |
+| $K$ - change constant | TODO |
+
+The initial parameters for the P-Chain will be set to:
+
+| Parameter | Value |
+| - | - |
+| $M$ - minimum gas price | TODO |
+| $T$ - target gas per second | TODO |
+| $L$ - max gas per second | TODO |
+| $K$ - change constant | TODO |
+
+As the network gains capacity to handle additional load (through the activation of future protocol enhancements), this algorithm can be tuned to increase the number of transactions that can be processed at a specific fee rate.
+
+#### A note on $e^x$
+
+There is a subtle reason why an exponential adjustment function was chosen: The adjustment function should be _equally_ reactive irrespective of the actual fee.
+
+Define $b_n$ as the current block's required fee, $b_{n+1}$ as the next block's required fee, and $x$ as the excess gas usage.
+
+Let's use a linear adjustment function:
+
+$$b_{n+1} = b_n + 10x$$
+
+Assume $b_n = 100$ and the current block is 1 unit above target utilization, or $x = 1$. Then, $b_{n+1} = 100 + 10 \cdot 1 = 110$, an increase of `10%`. If instead $b_n = 10,000$, $b_{n+1} = 10,000 + 10 \cdot 1 = 10,010$, an increase of `0.00001%`. The fee is _less_ reactive as the fee increases. This is because the rate of change is constant: $\frac{d}{dx}10x = 10$.
+
+Now, let's use an exponential adjustment function:
+
+$$b_{n+1} = b_n \cdot e^x$$
```

Comment:
q: wouldn't any multiplicative rule work here?
$$b_{n+1} = b_n \cdot f(x)$$

Reply:
$e$'s the simplest exponential function for theory purposes. We'll likely be using $2^x$ in the code

Reply:
The question was: why not using something that is not exponential? Say:
```
feeRate_{n+1} = feeRate_n \times x^2
```
You'd have something "equally reactive" as well.
I don't think we have explained why we need the exponential here